### DISCUSSION TITLE
Get Value from a IOEither example

### DISCUSSION THREAD

**Discussion by seyfer** [inato/fp-ts-cheatsheet#11] [Q&A] [answered]
Hey!

If I use `getOrElse` on `IOEither` I have to provide `IO` type value as a Left result, cannot be just `null`.

What I do now, I convert `IOEither` to `TaskEither` with `fromIOEither` function and then just run and await for `TaskEither`.
But that makes the whole `pipe` async. Any other way to get a value from `IOEither` without async?

**Comment by Punie** [comment, marked as answer]
Hello 🙂 

First, sorry for the delay, I somehow completely missed your issue 😞 
I still hope you'll find the (late) answer useful.

If I understand correctly, you are trying to do the following:

```ts
pipe(
  someIOEither,
  IOEither.getOrElse(() => null),
  ...do something with the A | null from the IOEither
)
``` 

And because it doesn't work, you are going through the trouble of converting your `IOEither` to a `TaskEither` (I assume because you are more comfortable with the latter?) which makes your computation `async`.

The problem you are having is that `IOEither<E, A>` is just sugar for `IO<Either<E, A>>` and the `getOrElse` only applies to the inner `Either`, collapsing the whole thing into an `IO<A>` that you can then just run like any other IO.

I think you then have two options:

- use `getOrElseW` (the W is important as you extend the `A` with `null`) and then run the `IO`:
```ts
pipe(
  someIOEither,
  IOEither.getOrElseW(() => IO.of(null)), // <= notice that you have to provide null wrapped in an IO, hence the IO.of(...)
  ...do something with the resulting IO<A | null>,
)() // <= run the IO to get the result
```
- first run the IO and `getOrElse` the resulting `Either<E, A>`:
```ts
pipe(
  someIOEither(), // <= run the IO immediately
  Either.getOrElseW(() => null), // <= notice we use Either<E, A> instead of IOEither<E, A> as the IO is already performed
  ...do something with the resulting A | null,
)
```

I hope that makes sense, please don't hesitate to reach out again if that wasn't clear enough 🙂
